src/main/lpfConverter.ts

```diff
@@ -100,110 +101,110 @@ async function injectManifestToZip(audiobookPath: string, manifest: Buffer) {
-            const nameArray = Array.isArray(name) ? name : [name];
-            const titleObj = nameArray.reduce(
-                (pv: any, cv: any) =>
-                    cv.value && cv.language ? pv[`${cv.language}`] = `${cv.value}` : pv,
-                {});
-            if (ObjectKeys(titleObj).length) {
-                publication.Metadata.Title = titleObj;
+            const nameObjArray = (Array.isArray(name) ? name : [name]) as IW3cName[];
+            const nameMap = nameObjArray.reduce(
+                (pv, cv: IW3cName) =>
+                    cv.value && cv.language ? pv.set(`${cv.language}`, `${cv.value}`) : pv,
+                new Map<string, string>());
+            if (nameMap.size) {
+                publication.Metadata.Title = Object.fromEntries(nameMap);
             }
 
         }
     }
     {
-        const language = w3cManifest.inLanguage;
-        if (typeof language === "string") {
-            publication.Metadata.Language = [language];
+        const language = pop("inLanguage");
+        const langArray = (Array.isArray(language) ? language : [language]);
+        const langArrayFiltered = langArray.filter((l) => typeof l === "string") as string[];
+        if (langArrayFiltered.length) {
+            publication.Metadata.Language = langArrayFiltered;
         }
     }
     {
-        const raw = `${w3cManifest.datePublished || ""}`;
+        const raw = `${pop("datePublished") || ""}`;
         const date = moment(raw);
         if (date.isValid()) {
             publication.Metadata.PublicationDate = date.toDate();
         }
     }
     {
-        const raw = `${w3cManifest.dateModified || ""}`;
+        const raw = `${pop("dateModified") || ""}`;
         const date = moment(raw);
         if (date.isValid()) {
             publication.Metadata.Modified = date.toDate();
         }
     }
     {
-        const iso8601 = `${w3cManifest.duration || ""}`;
+        const iso8601 = `${pop("duration") || ""}`;
         const second = iso8601DurationsToSeconds(iso8601);
         if (second > 0) {
             publication.Metadata.Duration = second;
         }
     }
     {
-        const resources = w3cManifest.resources as JsonMap;
+        const resources = pop("resources") as JsonMap;
         const links = convertW3CpublicationLinksToReadiumManifestLink(resources);
         if (links.length) {
             publication.Links = links;
         }
     }
     {
-        const readingOrder = w3cManifest.readingOrder as JsonMap;
+        const readingOrder = pop("readingOrder") as JsonMap;
         const links = convertW3CpublicationLinksToReadiumManifestLink(readingOrder);
         if (links.length) {
             publication.Spine = links;
             // https://github.com/readium/r2-shared-js/blob/develop/src/models/publication.ts#L63
         }
     }
+    {
+
+        if (Object.keys(w3cManifest).length) {
+            publication.Metadata.AdditionalJSON = deepClone(w3cManifest);
```

Comment:
Ah! `w3cManifest` is *not* an immutable object, in fact keys are deleted from this simple JSON structure (which is simply parsed from string) inside this very function. As the object is created outside this function, in principle the code in the function body breaks encapsulation, however in practice `w3cManifest` is an implementation detail inside "private" code, so it does not matter. Consequently, there is no need to clone `w3cManifest`, it can simply be assigned: `publication.Metadata.AdditionalJSON = w3cManifest;`. I would therefore suggest dropping the "deep clone" ad-hoc implementation, and instead in the future use `lodash` or some other robust micro lib. Alternatively, where performance is not an issue, this works fine too in the JSON case (i.e. only primitive types with array and object composition): `const clone = JSON.parse(JSON.stringify(obj));` ;)

Reply:
ah i just read your message
OK let's leave the ship :)